### PR TITLE
HDDS-5364. EC: Adopt EC related utility from Hadoop source repository

### PR DESCRIPTION
**Pull request by elek**
## What changes were proposed in this pull request?

It seems only a few key classes are required from Apache Hadoop to be reused. The goal here is investigating the build process and check how they can be re-used.

One option is just copy the files with some minor updates (e.g. using `ConfigurationSource` interface instead of `o.a.hadoop.Configuration`) which would help to reduce dependencies.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-5364

## How was this patch tested?

Full CI

### CODE REVIEW

**Comment by elek** [comment]
@sodonnel @umamaheswararao Can you please review it?

**Comment by umamaheswararao** [comment]
Thanks @elek for the patch. Sorry for the late on this. I am thinking that, we should copy the code as it is in JIRA and then modify in next JIRA? SInce it's large patch it may be hard to follow the changes against hadoop code here.
Thoughts?

CC: @sodonnel @fapifta

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java [hunk 1]
```diff
@@ -75,6 +87,7 @@ public int getRequiredNodes() {
     return HddsProtos.ECReplicationConfig.newBuilder()
         .setData(data)
         .setParity(parity)
+        .setCodec(codec)
         .build();
   }
```
Comment (umamaheswararao) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java:93]: Should we consider codec also in equals?

File: hadoop-hdds/erasurecode/src/main/java/org/apache/ozone/erasurecode/ECChunk.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+ * limitations under the License.
+ */
+package org.apache.ozone.erasurecode;
+
+import java.nio.ByteBuffer;
+
+/**
+ * A wrapper for ByteBuffer or bytes array for an erasure code chunk.
+ */
+public class ECChunk {
```
Comment (umamaheswararao) [inline hadoop-hdds/erasurecode/src/main/java/org/apache/ozone/erasurecode/ECChunk.java:25]: Looks like ECChunk abstraction we used only in tests? I think it's ok to have it for now and later we can cleanup if we don't use it ( to avoid code modifications in this patch)

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java [hunk 3]
```diff
@@ -28,18 +28,29 @@
  * Replication configuration for EC replication.
  */
 public class ECReplicationConfig implements ReplicationConfig {
-  
+
   private static final Pattern STRING_FORMAT = Pattern.compile("(\\d+)-(\\d+)");
```
Comment (umamaheswararao) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java:32]: Now we should be able to configure with codec option right? Could you please add comment to show the expected valid format ?

File: hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java [hunk 4]
```diff
@@ -28,18 +28,29 @@
...
 
+  public ECReplicationConfig(int data, int parity, String codec) {
+    this.data = data;
+    this.parity = parity;
+    this.codec = codec;
+  }
+
   public ECReplicationConfig(String string) {
     final Matcher matcher = STRING_FORMAT.matcher(string);
     if (!matcher.matches()) {
```
Comment (umamaheswararao) [inline hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/client/ECReplicationConfig.java:56]: You may want to update this message as we may allow codec also in config?

**Comment by umamaheswararao** [comment]
> Thanks the answer @umamaheswararao
> 
> However, I am not sure how is it possible. Just copying the code wouldn't compile. We have different dependency structure, package names and configuration objects.
> 
> But here I didn't change any logic, just copied the code and updated the used configuration interfaces. So technicaly this patch is the copy without any significant change.

I agree there is no logic change. Only worry is about the patch size and it's easy to miss things. For example, above conflict changes are not necessarily in this patch I guess. They are about using the new EcReplicationConfig with codec option. Currently they are already have default values and we can easily use this patch ECReplicationConfig in next patch. That way we can reduce modifications in this patch.

**Comment by umamaheswararao** [comment]
@elek did you get chance to check my comments? Thanks
